File: word-search-ii.cc
// Word Search II
#define REP(i, n) FOR(i, 0, n)
#define FOR(i, a, b) for (int i = (a); i < (b); i++)

class Solution {
  class TrieNode {
  public:
    int id;
    TrieNode *c[26];
    TrieNode() : id(-1) {
      fill_n(c, 26, nullptr);
    }
  } *root;

  int m, n;
  vector<vector<char>> board;
  vector<string> words, ans;

public:
  Solution() : root(new TrieNode) {}

  vector<string> findWords(vector<vector<char>> &board, vector<string> &words) {
    m = board.size();
    n = m ? board[0].size() : 0;
    REP(i, words.size()) {
      auto x = root;
      for (auto c: words[i]) {
        if (! x->c[c-'a'])
          x->c[c-'a'] = new TrieNode;
        x = x->c[c-'a'];
      }
      x->id = i;
    }
    this->board = board;
    this->words = words;
    ans.clear();
    REP(i, m)
      REP(j, n)
        f(i, j, root);
    return ans;
  }

  void f(int i, int j, TrieNode *x) {
    if (board[i][j] == '.') return;
    x = x->c[board[i][j]-'a'];
    if (! x) return;
    if (x->id >= 0) {
      ans.push_back(words[x->id]);
      x->id = -1;
    }
    char c = board[i][j];
    board[i][j] = '.';
    if (i) f(i-1, j, x);
    if (j+1 < n) f(i, j+1, x);
    if (i+1 < m) f(i+1, j, x);
    if (j) f(i, j-1, x);
    board[i][j] = c;
  }
};

File: word-search-ii.rs
#[derive(Default)]
struct Trie { c: [Option<Box<Self>>; 26], num: usize, w: Option<String> }

impl Solution {
    pub fn find_words(mut board: Vec<Vec<char>>, words: Vec<String>) -> Vec<String> {
        let mut trie: Trie = Default::default();
        for w in words.into_iter() {
            let mut x = &mut trie;
            for c in w.as_bytes().iter().map(|ch| (ch-b'a') as usize) {
                x = x.c[c].get_or_insert_with(|| Default::default());
                x.num += 1;
            }
            x.w = Some(w)
        }

        fn dfs(b: &mut Vec<Vec<char>>, ans: &mut Vec<String>, mut t: &mut Trie,
               x: usize, y: usize) -> usize {
            let ch = b[x][y];
            if ch == '.' { return 0; }
            if let Some(t1) = t.c[(ch as u8-b'a') as usize] {
                t = &mut *t1;
                if t.num == 0 { return 0; }
                let mut num = 0;
                if let Some(w) = t.w.take() {
                    ans.push(w);
                    num += 1;
                }
                b[x][y] = '.';
                if x > 0 { num += dfs(b, ans, t, x-1, y); }
                if x+1 < b.len() { num += dfs(b, ans, t, x+1, y); }
                if y > 0 { num += dfs(b, ans, t, x, y-1); }
                if y+1 < b[0].len() { num += dfs(b, ans, t, x, y+1); }
                b[x][y] = ch;
                t.num -= num;
                return num;
            }
            0
        }

        let m = board.len();
        let n = board[0].len();
        let mut ans = vec![];
        for x in 0..m {
            for y in 0..n {
                dfs(&mut board, &mut ans, &mut trie, x, y); } }
        ans
    }
}
